docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,519 @@
+```
+...
+ecall_table.ecalls = __oe_ecalls_table;
+ecall_table.num_ecalls = __oe_ecalls_table_size;
+
+if (args.function_id >= ecall_table.num_ecalls)
+    OE_RAISE(OE_NOT_FOUND);
+
+func = ecall_table.ecalls[args.function_id];
+...
+
+// Call the function.
+func(
+    input_buffer,
+    args.input_buffer_size,
+    output_buffer,
+    args.output_buffer_size,
+    &output_bytes_written);
+```
+
+After looking up based on the function id (`sample_fcn_id_sample_ecall`),
+`oe_handle_call_enclave_function()` calls into the wrapper function `ecall_sample_ecall()`(defined in `samlpe_t.c`).
+The wrapper function performs a series of checks, unmarshals the arguments, and invokes
+the ECALL implemented by the enclave with the following code snippet.
+
+```
+/* Call user function. */
+pargs_out->_retval = sample_ecall(
+);
+```
+
+At this point, an ECALL is successfully dispatched to the enclave.
+
+**OCALL Invocation**
+
+The invocation of OCALLs is similar to that of ECALLs. The only difference is that
+the array of the function pointers is defined as `__sample_ocall_function_table`
+and is kept as part of the `enclave` struct.
+More specifically, `sample_u.c` defines a wrapper function `oe_create_sample_enclave()` as follows.
+
+```
+oe_result_t oe_create_core_enclave(
+    const char* path,
+    oe_enclave_type_t type,
+    uint32_t flags,
+    const oe_enclave_setting_t* settings,
+    uint32_t setting_count,
+    oe_enclave_t** enclave)
+{
+    return oe_create_enclave(
+               path,
+               type,
+               flags,
+               settings,
+               setting_count,
+               __sample_ocall_function_table,
+               1,
+               enclave);
+}
+```
+
+The wrapper function is used to create the enclave by the host. The `__sample_ocall_function_table`
+is used during the initialization of the `enclave` struct as follows.
+
+```
+enclave->ocalls = (const oe_ocall_func_t*)ocall_table;
+enclave->num_ocalls = ocall_count;
+```
+
+Similarly, the `oe_handle_call_host_function()` function uses these variables to handle the requests
+of OCALLs.
+
+```
+...
+ocall_table.ocalls = enclave->ocalls;
+ocall_table.num_ocalls = enclave->num_ocalls;
+
+if (args_ptr->function_id >= ocall_table.num_ocalls)
+    OE_RAISE(OE_NOT_FOUND);
+
+func = ocall_table.ocalls[args_ptr->function_id];
+...
+
+// Call the function.
+func(
+    args_ptr->input_buffer,
+    args_ptr->input_buffer_size,
+    args_ptr->output_buffer,
+    args_ptr->output_buffer_size,
+    &args_ptr->output_bytes_written);
+```
+
+Limitations
+---------
+
+The current implementation poses the following limitations.
+
+- Single enclave cannot include multiple EDL files.
+
+  According to the implementation, the `oecore` library
+directly links against the `__oe_ecalls_table` defined in a `_t.c` file.
+Therefore, the current logic cannot handle the case of including
+multiple `_t.c` files.
+
+- Multiple enclaves on the same host cannot use/import the same EDL file.
```

Comment:
Consider rewording this more accurately. Multiple enclaves instances can today be created from the same enclave binary.

Reply:
The would no longer be true after introducing opt-in system EDLs. I'd like @jhand2's input here. He mentioned a case that two enclave import the same EDL may result in a conflict.

Reply:
Does this proposal solve the problem for hosts with multiple enclaves using overlapping sets of EDL files? i.e. is the only blocker today the function tables or do duplicate struct and type definitions prevent this as well?

Reply:
As I mentioned in another thread, the problem of addressing overlapping sets of EDL files is the main reason for this proposal to registering each function by hash. And this is why the existing implementation of per-EDL table is insufficient.